README.md
Make vanishing point boxes using HTML canvas


CHANGELOG.md
Commits

1. Plain simple box

![](https://i.imgur.com/sdIzbRR.png)

2. Back to start point

![](https://i.imgur.com/1ycrMzz.png)

3. added blue color to stroke

![](https://i.imgur.com/Z5VJ6JU.png)

4. Converting things from absolute pixel to variables + colorStrokes

![](https://i.imgur.com/N7eIygq.png)

5. Made everything a variable, can change delta

![](https://i.imgur.com/XvXA8YJ.png)

Code sample

![](https://i.imgur.com/Zk6JyQk.png)

6. Semi working slider

![](https://i.imgur.com/3Dzzojg.gif)

this is what it should do though??

![](https://i.imgur.com/BLFQngZ.png)

7. Working Prototype from stackoverflow, number coercion from slider

![](https://i.imgur.com/XU4AozG.gif)

8. Created partial line using slope formulas in purple

![](https://i.imgur.com/Yo1AwTF.png)

9. Working shaded value face

![](https://i.imgur.com/o6ykolr.png)

10. Simple box finished

![](https://i.imgur.com/q8ILYqV.png)

11. Growing Box

![](https://i.imgur.com/qr0Tb2E.gif)

12. Made a full house

![](https://i.imgur.com/6xFKNRG.gif)

13. Second Slider on baseY point does weird stuff

![](https://i.imgur.com/BRCHXiP.gif)

14. Slope_B now console.logs correctly

![](https://i.imgur.com/yAXKCR7.gif)

15. Had Slope formula upsidedown, swap x and y

![](https://i.imgur.com/QhUdwIq.gif)

16. Refactoring, part 1 destructuring with spread operator

![](https://i.imgur.com/JnxL44Q.png)

17. I made it work, two faces with remade formulas. Codes way cleaner now, code golf

![](https://i.imgur.com/8LxmBuh.png)

![](https://i.imgur.com/6LXc7NM.png)

18. this doesn't make sense. frontface=backface values, but canvas says otherwise

![](https://i.imgur.com/887DWnI.png)

![](https://i.imgur.com/bJS5UTi.png)

19. It must be an async thing. Console.logs are firing first

![](https://i.imgur.com/DsQHmTt.png)

20. Can specify which faces to write, for canvas overlap

![](https://i.imgur.com/TmAmaU8.gif)

21. Changeable origin sliders, but weird physics behavior

![](https://i.imgur.com/6cMk9SZ.gif)

22. Working Box and Origin X and Y

https://math.stackexchange.com/questions/175896/finding-a-point-along-a-line-a-certain-distance-away-from-another-point

As of now, everything is based on origin (0,0), (400,400) and (800,800)

Fixed it with a math formula from stackexchange

![](https://i.imgur.com/DGI5lsk.gif)

23. Added toggle fill color support

![](https://i.imgur.com/vf3Xo4Q.gif)

24. Adjustable Spacing Vanishing Point 2nd Line

![](https://i.imgur.com/CnnArWJ.gif)

25. Working Demo, but looks really off. Hard coded in 800 point offset

![](https://i.imgur.com/NreGGGO.gif)

26. Some issues with getting the box to be regular planed

![](https://i.imgur.com/F0okKQO.gif)


script-v2.js
// var sliders = document.querySelectorAll("input");
// sliders.forEach(item => {
//   item.addEventListener('input', (e) => {
//     item.nextElementSibling.textContent = e.target.value;
//     console.log(item.getAttribute("data-tag"));
//     window[item.getAttribute("data-tag")] = Number(e.target.value); // target variable by data-tag of input var name
//     redraw = true; // uses semaphore (a state variable) to indicate when to redraw
//   });
// })
// Init
requestAnimationFrame(update);
// Sliders
var canvas = document.getElementById("canvas");
var sliderSize = document.getElementById("sliderSize");
var sliderOriginY = document.getElementById("sliderOriginY");
var sliderOriginX = document.getElementById("sliderOriginX");
var sliderToggleFill = document.getElementById("sliderToggleFill");
sliderSize.addEventListener("input", function (e) {
    delta = Number((e.target.value));
    sliderSize.nextElementSibling.textContent = e.target.value;
    redraw = true; // uses semaphore (a state variable) to indicate when to redraw
});
sliderOriginY.addEventListener("input", function (e) {
    p1.y = Number((e.target.value));
    sliderOriginY.nextElementSibling.textContent = e.target.value;
    redraw = true; // uses semaphore (a state variable) to indicate when to redraw
});
sliderOriginX.addEventListener("input", function (e) {
    p1.x = Number((e.target.value));
    sliderOriginX.nextElementSibling.textContent = e.target.value;
    redraw = true; // uses semaphore (a state variable) to indicate when to redraw
});
sliderToggleFill.addEventListener("input", function (e) {
    toggleFill = Number((e.target.value));
    sliderToggleFill.nextElementSibling.textContent = e.target.value;
    redraw = true; // uses semaphore (a state variable) to indicate when to redraw
});
var ct = canvas.getContext("2d");
var width = 1600;
var height = 800;
canvas.width = window.innerWidth;
canvas.height = window.innerHeight;
// Scales 2D context to always show ideal resolution area
var scaleToFit = function () {
    var scale = Math.min(canvas.width / width, canvas.height / height);
    ct.setTransform(scale, 0, 0, scale, 0, 0); // makes canvas fit to screen
};
var redraw = true; // When true, scene is redrawn ready for the next display refresh
// Constructor Function to initialize objects
var point = function (x, y) {
    if (x === void 0) { x = 0; }
    if (y === void 0) { y = 0; }
    return ({ x: x, y: y });
};
var pointCalc = function (p, x, y) {
    if (x === void 0) { x = 0; }
    if (y === void 0) { y = 0; }
    return ({ x: p.x + x, y: p.y + y });
};
// y = mx + b slope formula for other points
var scalePoint = function (origin, point, scale) {
    point.x = (point.x - origin.x) * scale + origin.x;
    point.y = (point.y - origin.y) * scale + origin.y;
};
// Objects Cartesion Points
var p1 = point(400, 400);
;
var pA = point(p1.x, p1.y * 2);
var pB = point(p1.x * 2, p1.y * 2);
var delta = 50;
var toggleFill = 0;
function update() {
    if (redraw) { // monitor semaphore / state
        redraw = false; // clear semaphore / state
        ct.setTransform(1, 0, 0, 1, 0, 0); // resets transform
        ct.clearRect(0, 0, ct.canvas.width, ct.canvas.height); // clear Canvas
        scaleToFit();
        draw();
    }
    requestAnimationFrame(update);
}
// was function init(), specify start and endpoint
function draw() {
    drawLine(p1, pA, "red");
    drawLine(p1, pB, "green");
    drawVBox(pB, delta, p1, "blue", 1);
}
function drawVBox(p, size, vp, col, width) {
    ct.strokeStyle = col;
    ct.lineWidth = width;
    0;
    var frontFace = [
        pointCalc(p),
        pointCalc(p, size, 0),
        pointCalc(p, size, -size),
        pointCalc(p, 0, -size)
    ];
    console.log(frontFace, "frontFace");
    var scale = 1 - size / (800 * 2); // Inverse scalar
    console.log(scale);
    //https://math.stackexchange.com/questions/175896/finding-a-point-along-a-line-a-certain-distance-away-from-another-point
    var backFace = frontFace.map(function (_a) {
        var x = _a.x, y = _a.y;
        return ({
            x: vp.x - scale * (vp.x - x),
            y: vp.y - scale * (vp.y - y)
        });
    });
    // bottomleft, bottomright, topright, topleft
    drawPoly.apply(void 0, [col, width].concat(backFace)); // back
    drawPoly(col, width, backFace[0], backFace[3], frontFace[3], frontFace[0]); // left
    drawPoly(col, width, backFace[1], backFace[2], frontFace[2], frontFace[1]); // right
    drawPoly.apply(void 0, [col, width].concat(frontFace)); // front
}
// sideface
function makeFaceYZ(p, vp) {
}
// front / backface
function makeFaceXZ(p, vp) {
}
// topface / bottomface
function makeFaceXY(p, vp) {
}
function calculateSlope(p, vp) {
    return (p.y - vp.y) / (p.x - vp.x);
}
// Use function to do common tasks and save yourself lots of typing
function drawLine(p1, p2, col, width) {
    if (width === void 0) { width = 1; }
    // ct.globalCompositeOperation = 'destination-over'; // draw behind
    ct.strokeStyle = col;
    ct.lineWidth = width;
    ct.beginPath();
    ct.lineTo(p1.x, p1.y); // First Point after beginPath can be lineTo (don't need moveTo)
    ct.lineTo(p2.x, p2.y);
    ct.stroke();
}
// For vanishing lines only
function pathLine(p1, p2) {
    ct.moveTo(p1.x, p1.y);
    ct.lineTo(p2.x, p2.y);
}
// Draw the square
function drawPoly(col, width) {
    var points = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        points[_i - 2] = arguments[_i];
    }
    ct.strokeStyle = col;
    ct.lineWidth = width;
    if (toggleFill == 1) {
        ct.fillStyle = 'lightgreen';
    }
    else {
        ct.fillStyle = "rgba(0, 0, 200, 0)";
    }
    ct.beginPath();
    for (var _a = 0, points_1 = points; _a < points_1.length; _a++) {
        var p = points_1[_a];
        ct.lineTo(p.x, p.y); // lineTo every point
    }
    ct.fill();
    ct.closePath(); // draw closing line
    ct.stroke();
}


script-v2.ts
// var sliders = document.querySelectorAll("input");
// sliders.forEach(item => {
//   item.addEventListener('input', (e) => {
//     item.nextElementSibling.textContent = e.target.value;
//     console.log(item.getAttribute("data-tag"));
//     window[item.getAttribute("data-tag")] = Number(e.target.value); // target variable by data-tag of input var name
//     redraw = true; // uses semaphore (a state variable) to indicate when to redraw
//   });
// })

// Init
requestAnimationFrame(update);

// Sliders
var canvas: any = document.getElementById("canvas");
var sliderSize = document.getElementById("sliderSize");
var sliderOriginY = document.getElementById("sliderOriginY");
var sliderOriginX = document.getElementById("sliderOriginX");
var sliderToggleFill = document.getElementById("sliderToggleFill");
sliderSize.addEventListener("input", (e) => {
  delta = Number((e.target.value);
  sliderSize.nextElementSibling.textContent = e.target.value;
  redraw = true; // uses semaphore (a state variable) to indicate when to redraw
})
sliderOriginY.addEventListener("input", (e) => {
  p1.y = Number((e.target.value);
  sliderOriginY.nextElementSibling.textContent = e.target.value;
  redraw = true; // uses semaphore (a state variable) to indicate when to redraw
})
sliderOriginX.addEventListener("input", (e) => {
  p1.x = Number((e.target.value);
  sliderOriginX.nextElementSibling.textContent = e.target.value;
  redraw = true; // uses semaphore (a state variable) to indicate when to redraw
})
sliderToggleFill.addEventListener("input", (e) => {
  toggleFill = Number((e.target.value);
  sliderToggleFill.nextElementSibling.textContent = e.target.value;
  redraw = true; // uses semaphore (a state variable) to indicate when to redraw
})



const ct = canvas.getContext("2d");
const width = 1600;
const height = 800;
canvas.width = window.innerWidth;
canvas.height = window.innerHeight;

// Scales 2D context to always show ideal resolution area
const scaleToFit = () => {
  var scale = Math.min(canvas.width / width, canvas.height / height);
  ct.setTransform(scale, 0, 0, scale, 0, 0); // makes canvas fit to screen
}

var redraw = true; // When true, scene is redrawn ready for the next display refresh

// Constructor Function to initialize objects
const point = (x = 0, y = 0) => ({ x, y });
const pointCalc = (p, x = 0, y = 0) => ({ x: p.x + x, y: p.y + y });
// y = mx + b slope formula for other points
const scalePoint = (origin, point, scale) => {
  point.x = (point.x - origin.x) * scale + origin.x;
  point.y = (point.y - origin.y) * scale + origin.y;
};

// Objects Cartesion Points
var p1 = point(400, 400);;
var pA = point(p1.x, p1.y * 2);
var pB = point(p1.x * 2, p1.y * 2);

var delta = 50;
var toggleFill = 0;

function update() { // render loop redraws only if true
  if (redraw) { // monitor semaphore / state
    redraw = false; // clear semaphore / state
    ct.setTransform(1, 0, 0, 1, 0, 0); // resets transform
    ct.clearRect(0, 0, ct.canvas.width, ct.canvas.height); // clear Canvas
    scaleToFit();
    draw();
  }
  requestAnimationFrame(update);
}

// was function init(), specify start and endpoint
function draw() {
  drawLine(p1, pA, "red");
  drawLine(p1, pB, "green");
  drawVBox(pB, delta, p1, "blue", 1);
}

function drawVBox(p, size, vp, col, width) { // p is bottom left,  vp is vanish point
  ct.strokeStyle = col;
  ct.lineWidth = width; 0

  const frontFace = [
    pointCalc(p),
    pointCalc(p, size, 0),
    pointCalc(p, size, -size),
    pointCalc(p, 0, -size)
  ]
  console.log(frontFace, "frontFace");
  const scale = 1 - size / (800 * 2); // Inverse scalar
  console.log(scale);

  //https://math.stackexchange.com/questions/175896/finding-a-point-along-a-line-a-certain-distance-away-from-another-point
  const backFace = frontFace.map(({ x, y }) => ({
    x: vp.x - scale * (vp.x - x),
    y: vp.y - scale * (vp.y - y)
  }));
  // bottomleft, bottomright, topright, topleft
  drawPoly(col, width, ...backFace); // back
  drawPoly(col, width, backFace[0], backFace[3], frontFace[3], frontFace[0]); // left
  drawPoly(col, width, backFace[1], backFace[2], frontFace[2], frontFace[1]); // right
  drawPoly(col, width, ...frontFace); // front
}

// sideface
function makeFaceYZ(p, vp) {
}
// front / backface
function makeFaceXZ(p, vp) {

}
// topface / bottomface
function makeFaceXY(p, vp) {

}

function calculateSlope(p, vp) {
  return (p.y - vp.y) / (p.x - vp.x);
}

// Use function to do common tasks and save yourself lots of typing
function drawLine(p1, p2, col, width = 1) {
  // ct.globalCompositeOperation = 'destination-over'; // draw behind
  ct.strokeStyle = col;
  ct.lineWidth = width;
  ct.beginPath();
  ct.lineTo(p1.x, p1.y); // First Point after beginPath can be lineTo (don't need moveTo)
  ct.lineTo(p2.x, p2.y);
  ct.stroke();
}

// For vanishing lines only
function pathLine(p1, p2) {
  ct.moveTo(p1.x, p1.y);
  ct.lineTo(p2.x, p2.y);
}

// Draw the square
function drawPoly(col, width, ...points) {
  ct.strokeStyle = col;
  ct.lineWidth = width;
  if (toggleFill == 1) {
    ct.fillStyle = 'lightgreen';
  } else {
    ct.fillStyle = "rgba(0, 0, 200, 0)";
  }
  ct.beginPath();
  for (const p of points) {
    ct.lineTo(p.x, p.y); // lineTo every point
  }

  ct.fill();
  ct.closePath(); // draw closing line
  ct.stroke();
}


version1.0.1/script-v1.js
// for reference https://codepen.io/AshKyd/pen/JYXEpL, html5 canvas isometric cube

var canvas = document.querySelector("canvas");
canvas.width = window.innerWidth;
canvas.height = window.innerHeight;
canvas.width = 1600;
canvas.height = 800;

var ct = canvas.getContext("2d");
var x0 = 400;
var y0 = 400;

// Vanishing point end 0a
var x0a = 0;
// var y0a = 2 * y0;
var y0a = 800;

// Vanishing point end 0b
// var x0b = 2 * x0;
// var y0b = 2 * y0;
var x0b = 800;
var y0b = 800;

// Define length of box edges
var size = 100;

function init() {
  // First Line
  ct.beginPath();
  ct.moveTo(x0, y0);
  ct.lineTo(x0a, y0a);
  ct.strokeStyle = 'red';
  ct.stroke();

  // // Second Line
  ct.beginPath();
  ct.moveTo(x0, y0);
  ct.lineTo(x0b, x0b);
  ct.strokeStyle = 'green';
  ct.stroke();

  // Draw a bunch of houses
  drawHouse(x0b, y0b, size);
}

function shadeColor() {

}

function drawHouse(startX, startY, size, color) {
  // House based on second Line

  // Formula slopes
  // m = (y1 - y0)/(x1 - x0)
  // y1 = mx1 - mx0 + y0

  // Line distance Formula
  // d = sqrt((x1-x0)^2 + (y1-y0)^2)

  // Pythagoreans Theorm
  // c^2 = a^2 + b^2

  // Height of a 45* triangle
  // Area = 1/2*base*height
  // height = size variable

  // Go here "\"
  // Nomenclature
  // a1,b1 = one coordinate pair
  // a2,b2 = another coordinate pair
  // go counter clockwise on the points

  // calculate same sized lined box

  // Calculate the 2nd point on vanishing line
  var slope_b = (y0b - y0) / (x0b - x0);
  let a1 = startX - size / 1.9; // ~45* angle approximation to make sure all lengths of square are same size. Use line distance / pythagoreans theorm for accuracy
  let b1 = slope_b * (a1 - x0) + y0

  console.log(x0, "x0");
  console.log(y0, "y0");
  console.log(slope_b, "slope_b");
  console.log(x0b, "x0b");
  console.log(y0b, "y0b");
  console.log(a1, "a1");
  console.log(b1, "b1");


  // Side Face
  ct.beginPath();
  ct.moveTo(startX, startY);
  ct.lineTo(a1, b1);
  ct.lineTo(a1, b1 - size);
  ct.lineTo(startX, startY - size);
  ct.closePath();
  ct.strokeStyle = 'purple';
  ct.fillStyle = 'red';
  ct.stroke();
  ct.fill();

  // Front Face
  ct.beginPath();
  ct.moveTo(startX, startY);
  ct.lineTo(startX + size, startY);
  ct.lineTo(startX + size, startY - size);
  ct.lineTo(startX, startY - size);
  ct.closePath();
  ct.strokeStyle = 'purple';
  ct.fillStyle = 'lightblue';
  ct.stroke();
  ct.fill();

  // Top Face - Not necessary but whatever
  ct.beginPath();
  ct.moveTo(startX, startY - size);
  ct.lineTo(a1, b1 - size);
  ct.lineTo(a1 + size, b1 - size);
  ct.lineTo(startX + size, startY - size);
  ct.closePath();
  ct.strokeStyle = 'purple';
  ct.fillStyle = 'lightgreen';
  ct.stroke();
  ct.fill();

  // Draw Roof Back Face - Not necessary but whatever
  ct.beginPath();
  ct.moveTo(a1, b1 - size);
  ct.lineTo(a1 + size / 2, b1 - 2 * size);
  ct.lineTo(a1 + size, b1 - size);
  ct.closePath();
  ct.strokeStyle = 'purple';
  ct.fillStyle = 'pink';
  ct.stroke();
  ct.fill();

  // Draw a Roof Front Face
  ct.beginPath();
  ct.moveTo(startX, startY - size);
  ct.lineTo(startX + size / 2, startY - 2 * size);
  ct.lineTo(startX + size, startY - size);
  ct.closePath();
  ct.strokeStyle = 'purple';
  ct.fillStyle = 'pink';
  ct.stroke();
  ct.fill();

  // Shade frontside
  ct.beginPath();
  ct.moveTo(startX, startY - size);
  ct.lineTo(a1, b1 - size);
  ct.lineTo(a1 + size / 2, b1 - 2 * size);
  ct.lineTo(startX + size / 2, startY - 2 * size);
  ct.closePath();
  ct.strokeStyle = 'purple';
  ct.fillStyle = 'lightblue';
  ct.stroke();
  ct.fill();

}


init();


// Slider Functionality
var slider = document.getElementById("mySize");
var slider2 = document.getElementById("myBaseY");
var slider3 = document.getElementById("myBaseX");
slider.oninput = function () {
  ct.clearRect(0, 0, canvas.width, canvas.height); // reset lines
  size = +this.value; // convert the slider value to a number for type coersion (see stackoverflow)
  requestAnimationFrame(init()); // redraw everything
}

slider2.oninput = function () {
  ct.clearRect(0, 0, canvas.width, canvas.height); // reset lines
  y0 = Number(this.value); // convert the slider value to a number for type coersion (see stackoverflow)
  requestAnimationFrame(init()); // redraw everything
}
slider3.oninput = function () {
  ct.clearRect(0, 0, canvas.width, canvas.height); // reset lines
  x0 = Number(this.value); // convert the slider value to a number for type coersion (see stackoverflow)
  requestAnimationFrame(init()); // redraw everything
}

script.js
requestAnimationFrame(update);

// Sliders
var canvas = document.getElementById("canvas");
var sliderSize = document.getElementById("sliderSize");
var sliderOriginY = document.getElementById("sliderOriginY");
var sliderOriginX = document.getElementById("sliderOriginX");
var sliderToggleFill = document.getElementById("sliderToggleFill");
var sliderSpacing = document.getElementById("sliderSpacing");
var sliderElongate = document.getElementById("sliderElongate");
// Size
sliderSize.addEventListener("input", e => {
  delta = Number(e.target.value);
  sliderSize.nextElementSibling.textContent = e.target.value;
  redraw = true;
});
// OriginY
sliderOriginY.addEventListener("input", e => {
  p1.y = Number(e.target.value);
  q1.y = Number(e.target.value); // keep y value on same plane
  sliderOriginY.nextElementSibling.textContent = e.target.value;
  redraw = true;
});
// OriginX
sliderOriginX.addEventListener("input", e => {
  p1.x = Number(e.target.value);
  q1.x = Number(e.target.value) * spacing;
  console.log(spacing, "spacing");

  sliderOriginX.nextElementSibling.textContent = e.target.value;
  redraw = true;
});
//Toggle Fill
sliderToggleFill.addEventListener("input", e => {
  toggleFill = Number(e.target.value);
  sliderToggleFill.nextElementSibling.textContent = e.target.value;
  redraw = true;
});
//Toggle SecondVP
sliderAddVP.addEventListener("input", e => {
  secondVP = Number(e.target.value);
  sliderAddVP.nextElementSibling.textContent = e.target.value;
  redraw = true;
});
// Toggle Spacing Distance
sliderSpacing.addEventListener("input", e => {
  spacing = 1 + Number(e.target.value) / 100;
  p1.x = Number(e.target.value);
  q1.x = Number(e.target.value) * spacing;
  sliderSpacing.nextElementSibling.textContent = e.target.value;
  redraw = true;
});
// Toggle Rectangilness
sliderElongate.addEventListener("input", e => {
  scalar = Number(e.target.value);
  sliderElongate.nextElementSibling.textContent = e.target.value;
  redraw = true;
});

// Document Sliders

// var msgBox = document.querySelector(".checkbox");
// msgBox.addEventListener("input", e => {
//   console.log(e.target.value);
// });
// console.log(msgBox);

const ct = canvas.getContext("2d");
const width = 1600;
const height = 800;
canvas.width = window.innerWidth;
canvas.height = window.innerHeight;

// Scales 2D context to always show ideal resolution area
const scaleToFit = () => {
  var scale = Math.min(canvas.width / width, canvas.height / height);
  ct.setTransform(scale, 0, 0, scale, 0, 0); // makes canvas fit to screen
};

var redraw = true; // When true, scene is redrawn ready for the next display refresh

// Constructor Function to initialize objects
const point = (x = 0, y = 0) => ({ x, y });
const pointCalc = (p, x = 0, y = 0) => ({ x: p.x + x, y: p.y + y });
// y = mx + b slope formula for other points
const scalePoint = (origin, point, scale) => {
  point.x = (point.x - origin.x) * scale + origin.x;
  point.y = (point.y - origin.y) * scale + origin.y;
};

// Objects Cartesion Points
var p1 = point(400, 400);
var pA = point(p1.x, p1.y * 2);
var pB = point(p1.x * 2, p1.y * 2);

// Second Vanishing Point
var q1 = point(p1.x + 800, p1.y);

// State Toggles
var delta = 50;
var toggleFill = 0;
var secondVP = 0;
var spacing = 1.15;
var scalar = 1600;

function update() {
  // render loop redraws only if true
  if (redraw) {
    // monitor semaphore / state
    redraw = false; // clear semaphore / state
    ct.setTransform(1, 0, 0, 1, 0, 0); // resets transform
    ct.clearRect(0, 0, ct.canvas.width, ct.canvas.height); // clear Canvas
    scaleToFit();
    draw();
  }
  requestAnimationFrame(update);
}

// was function init(), specify start and endpoint
function draw() {
  drawLine(p1, pA, "red");
  drawLine(p1, pB, "green");
  drawVBox(pB, delta, p1, "blue", 1, q1, scalar);
  if (secondVP == 1) {
    drawLine(q1, pB, "purple");
  }
}

function drawVBox(p, size, vp, col, width, vp2, scalar) {
  //https://math.stackexchange.com/questions/175896/finding-a-point-along-a-line-a-certain-distance-away-from-another-point
  // p is bottom left,  vp is vanish point
  ct.strokeStyle = col;
  ct.lineWidth = width;
  // var elongate = 1 - size / (800 * 2); // Make box more elongated
  var elongate = 1 - size / scalar; // Make box more elongated
  let frontFace = [];
  let backFace = [];

  if (secondVP == 0) {
    frontFace = [
      pointCalc(p),
      pointCalc(p, size, 0),
      pointCalc(p, size, -size),
      pointCalc(p, 0, -size)
    ];
    backFace = frontFace.map(({ x, y }) => ({
      x: vp.x - elongate * (vp.x - x),
      y: vp.y - elongate * (vp.y - y)
    }));
    console.log(backFace, "backface, 1st VP");
    console.log(frontFace, "frontface, 1st VP");
  } else if (secondVP == 1) {
    frontFace = [
      pointCalc(p),
      pointCalc(
        p,
        vp2.x - elongate * (vp2.x - p.x) - 800,
        vp2.y - elongate * (vp.y - p.y) - 800
      ),
      pointCalc(
        p,
        vp2.x - elongate * (vp2.x - p.x) - 800,
        vp2.y - elongate * (vp.y - p.y) - 800 - size
      ),
      pointCalc(p, 0, -size)
    ];
    // backFace = frontFace.map(({ x, y }) => ({
    //   x: vp.x - elongate * (vp.x - x),
    //   y: vp.y - elongate * (vp.y - y)
    // }));
    backFace = frontFace.map(({ x, y }) => ({
      x: vp.x - elongate * (vp.x - x),
      y: vp.y - elongate * (vp.y - y)
    }));
    console.log(backFace, "backface, 2nd VP");
    console.log(frontFace, "frontface, 2nd VP");
  } else {
    console.log(
      "error, can only have 1 or 2 vanishing points when establishing coordinates"
    );
  }

  // If only one VP
  if (secondVP == 0 || secondVP == 1) {
    // bottomleft, bottomright, topright, topleft
    drawPoly(col, width, ...backFace); // back
    drawPoly(col, width, backFace[0], backFace[3], frontFace[3], frontFace[0]); // left
    drawPoly(col, width, backFace[1], backFace[2], frontFace[2], frontFace[1]); // right
    drawPoly(col, width, ...frontFace); // front
  } else {
    console.log("error, can only have 1 or 2 vanishing points when drawing");
  }
}

// Use function to do common tasks and save yourself lots of typing
function drawLine(p1, p2, col, width = 1) {
  // ct.globalCompositeOperation = 'destination-over'; // draw behind
  ct.strokeStyle = col;
  ct.lineWidth = width;
  ct.beginPath();
  ct.lineTo(p1.x, p1.y); // First Point after beginPath can be lineTo (don't need moveTo)
  ct.lineTo(p2.x, p2.y);
  ct.stroke();
}

// For vanishing lines only
function pathLine(p1, p2) {
  ct.moveTo(p1.x, p1.y);
  ct.lineTo(p2.x, p2.y);
}

// Draw the square
function drawPoly(col, width, ...points) {
  ct.strokeStyle = col;
  ct.lineWidth = width;
  if (toggleFill == 1) {
    ct.fillStyle = "lightgreen";
  } else {
    ct.fillStyle = "rgba(0, 0, 200, 0)";
  }
  ct.beginPath();
  for (const p of points) {
    ct.lineTo(p.x, p.y); // lineTo every point
  }

  ct.fill();
  ct.closePath(); // draw closing line
  ct.stroke();
}
